### PR TITLE
perform rollback when deployment fails

### LINKED ISSUE #787: perform rollback when deployment fails
As part of terraform apply, if any steps fails after some resources are created, then terraform does not do a automatic rollback. 

- [ ] After the deployment fails in between due to a error from Terraform plan/apply, store the already provisioned resources like we do in case of success cases https://github.com/eclipse-xpanse/xpanse/blob/27a366f138933b7395eee293c7950e9c049881e4/modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java#L196
- [ ] if any resources were already provisioned, execute a destroy immediately. . 
- [ ] Ensure the deployment failure message is correctly stored and not overwritten by destroy messages. 
- [ ] If the Destroy fails set the service to a new status MANUAL_CLEANUP_REQUIRED.
- [ ] service state remains as DEPLOY_FAILED if the destroy is successful.

### PR DESCRIPTION
**Pull request by jinyangyang222**
fixes #787

### CODE REVIEW

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 1]
```diff
@@ -202,10 +203,49 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
...
+                        ServiceDeploymentState.DEPLOY_FAILED);
+                deployServiceEntity.setProperties(destroyResult.getProperties());
+                deployServiceEntity.setPrivateProperties(destroyResult.getPrivateProperties());
+                deployServiceEntity.setDeployResourceList(
+                        getDeployResourceEntityList(destroyResult.getResources(),
+                                deployServiceEntity));
+                deployServiceStorage.storeAndFlush(deployServiceEntity);
+                List<DeployResource> resources = destroyResult.getResources();
+                if (CollectionUtils.isEmpty(resources)) {
+                    deployResourceStorage.deleteByDeployServiceId(deployServiceEntity.getId());
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:236]: there is no need to delete the service from DB.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 2]
```diff
@@ -202,10 +203,49 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
...
+     * Perform rollback when deployment fails and destroy the created resources.
+     */
+    private void rollbackOnDeploymentFailure(DeployServiceEntity deployServiceEntity,
+                                             DeployResult deployResult, Deployment deployment,
+                                             DeployTask deployTask) {
+        if (!CollectionUtils.isEmpty(deployResult.getResources())) {
+            String stateFile = deployServiceEntity.getPrivateProperties().get(STATE_FILE_NAME);
+            DeployResult destroyResult = deployment.destroy(deployTask, stateFile);
+            if (destroyResult.getState() == TerraformExecState.DESTROY_SUCCESS) {
+                deployServiceEntity.setServiceDeploymentState(
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:226]: this is already done in line 204?

**Review by swaroopar — CHANGES_REQUESTED** [review]
comments added.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 3]
```diff
@@ -202,10 +203,49 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
...
+                        ServiceDeploymentState.DEPLOY_FAILED);
+                deployServiceEntity.setProperties(destroyResult.getProperties());
+                deployServiceEntity.setPrivateProperties(destroyResult.getPrivateProperties());
+                deployServiceEntity.setDeployResourceList(
+                        getDeployResourceEntityList(destroyResult.getResources(),
+                                deployServiceEntity));
+                deployServiceStorage.storeAndFlush(deployServiceEntity);
```
[the remark this answers is not in the document]
Comment (jinyangyang222) [reply modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java]: fixed

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 4]
```diff
@@ -202,10 +203,39 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
...
+    }
 
+    /**
+     * Perform rollback when deployment fails and destroy the created resources.
+     */
+    private void rollbackOnDeploymentFailure(DeployServiceEntity deployServiceEntity,
+                                             DeployResult deployResult, Deployment deployment,
+                                             DeployTask deployTask) {
+        if (!CollectionUtils.isEmpty(deployResult.getResources())) {
+            String stateFile = deployServiceEntity.getPrivateProperties().get(STATE_FILE_NAME);
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:223]: please a log line -> "performing rollback of already provisioned resources"

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 5]
```diff
@@ -202,10 +203,39 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
...
+                        getDeployResourceEntityList(destroyResult.getResources(),
+                                deployServiceEntity));
+            } else {
+                deployServiceEntity.setServiceDeploymentState(
+                        ServiceDeploymentState.MANUAL_CLEANUP_REQUIRED);
+            }
+            if (deployServiceStorage.storeAndFlush(deployServiceEntity)) {
+                deployment.deleteTaskWorkspace(deployTask.getId().toString());
+            }
+        }
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:238]: please a log line for else block -> "No resources to rollback. "

Comment (jinyangyang222) [reply modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:239]: fixed

**Review by swaroopar — APPROVED** [review]
LGTM

**Pull request merged**